src/main/java/org/springframework/integration/kafka/core/DefaultConnectionFactory.java

```diff
@@ -0,0 +1,258 @@
+	 * throughput and increasing CPU consumption.
+	 *
+	 * @param minBytes
+	 */
+	public void setMinBytes(int minBytes) {
+		this.minBytes = minBytes;
+	}
+
+	public int getMaxWait() {
+		return maxWait;
+	}
+
+	/**
+	 * The maximum amount of time that a server fetch operation will wait before returning (unless {@code minFetchSizeInBytes})
+	 * are available. In conjunction with {@link DefaultConnectionFactory#setMinBytes(int)},
+	 * controls latency and throughput.
+	 *
+	 * Smaller intervals increase responsiveness, but may increase the number of poll operations, potentially increasing CPU
+	 * consumption and reducing throughput.
+	 *
+	 * @param maxWait
+	 */
+	public void setMaxWait(int maxWait) {
+		this.maxWait = maxWait;
+	}
+
+	public String getClientId() {
+		return clientId;
+	}
+
+	/**
+	 * A client name to be used throughout this connection.
+	 *
+	 * @param clientId
+	 */
+	public void setClientId(String clientId) {
+		this.clientId = clientId;
+	}
+
+	public int getBufferSize() {
+		return bufferSize;
+	}
+
+	/**
+	 * The buffer size for this client
+	 * @param bufferSize
+	 */
+	public void setBufferSize(int bufferSize) {
+		this.bufferSize = bufferSize;
+	}
+
+	public int getSocketTimeout() {
+		return socketTimeout;
+	}
+
+	/**
+	 * The socket timeout for this client
+	 * @param socketTimeout
+	 */
+	public void setSocketTimeout(int socketTimeout) {
+		this.socketTimeout = socketTimeout;
+	}
+
+	/**
+	 * @see ConnectionFactory#getLeaders(Iterable)
+	 */
+	@Override
+	public Map<Partition, BrokerAddress> getLeaders(Iterable<Partition> partitions) {
+		return FastList.newList(partitions).toMap(Functions.<Partition>getPassThru(), getBrokersByPartitionFunction);
+	}
+
+	/**
+	 * @see ConnectionFactory#getLeader(Partition)
+	 */
+	@Override
+	public BrokerAddress getLeader(Partition partition) {
+		try {
+			lock.readLock().lock();
+			return this.getLeaders(Collections.singleton(partition)).get(partition);
+		}
+		finally {
+			lock.readLock().unlock();
+		}
+	}
+
+	/**
+	 * @see ConnectionFactory#connect(BrokerAddress)
+	 */
+	@Override
+	public Connection connect(BrokerAddress brokerAddress) {
+		return kafkaBrokersCache.getIfAbsentPutWithKey(brokerAddress, connectionInstantiationFunction);
+	}
+
+	/**
+	 * @see ConnectionFactory#refreshLeaders(Collection)
+	 */
+	@Override
+	public void refreshLeaders(Collection<String> topics) {
+		try {
+			lock.writeLock().lock();
+			for (Connection connection : kafkaBrokersCache) {
+				connection.close();
+			}
+			String brokerAddressesAsString = ListIterate.collect(configuration.getBrokerAddresses(), Functions.getToString()).makeString(",");
+			TopicMetadataResponse topicMetadataResponse = new TopicMetadataResponse(ClientUtils$.MODULE$.fetchTopicMetadata(
+					JavaConversions.asScalaSet(new HashSet<String>(topics)), ClientUtils$.MODULE$.parseBrokerList(brokerAddressesAsString), getClientId(), 10000, 0));
```

Comment:
Should that `fetchTopicMetadata` timeout be configurable?
In your case it is `10000`